### DISCUSSION TITLE
[2023-06-05] GUI Team Demo Day

### DISCUSSION THREAD

**Discussion by wdanilo** [enso-org/enso#6946] [Weekly Demo Day]
On my end, I was working on new FRP implementation. We have a new, really fast working impl, that is incomparably faster to the old impl. The code is in its not-cleaned state and I plan to clean it till tomorrow. Unfortunately, it is really hard in many places and I needed to create a lot of abstractions and fix a lot of abstractions we had, but were implemented not in the best way. The code can be browsed [here](https://github.com/enso-org/enso/pull/6944), but I'd advise to wait till it is cleaned before reviewing it. The upcoming PR will contain the FRP implementation that will not replace the old one yet. The replacement will be done in another PR this week.

**Comment by Frizi** [comment]
- finished https://github.com/enso-org/enso/pull/6878, and with it closed multiple related issues
- Implementation of new edge disconnect behavior is still in progress https://github.com/enso-org/enso/issues/6772
  - When disconnecting the edge at target side, it is supposed to not edit the node expression until the edge is dropped/reconnected. That way we always have a valid edge target.
  - I'm in process of changing edge disconnection handling within graph editor FRP network, so that the new behavior can be implemented, and to make edges more resilient to expression updates in general.
- Discussed changes necessary for better mouse hover handling, planning to implement that as a next task. https://github.com/enso-org/enso/issues/6850#issuecomment-1576754037
- Supporting engine changes related to reported method pointers in expression updates https://github.com/orgs/enso-org/discussions/6762

**Reply by sylwiabr** [reply]
@Frizi lets push the changes for new design first and then move to #6850

**Comment by farmaazon** [comment]
My fixes:

**Disabled entering node when editing it**, because it sometimes interfered with selecting word by double click.

**Implemented proper handling of "partial success" when entering node.** As entering node or graph editor's breadcrumb require several calls to the engine, sometimes we faced a failure after some successful calls, leaving execution context in wrong state (that's why the infamous ["File not found error"](#6763) made engine looking broken). Now, the code was updated to handle carefully such situations.

**Still in progress: fixing HTML elements not getting keyboard events** (i.e. the issues like not-working filter field in the Table visualization, or not working "rename" field in dashboard after exiting project). Tested what happens after attaching event listeners to "scene" div instead of window - this indeed fixes the dashboard, but not visualizations. I will deliver it as first fix, I only try to find a way to "autofocus" it (now, after opening project, you need to click inside or press tab to have shortcuts working).

**Comment by vitvakatu** [comment]
Last week I was mainly working on updating the documentation panel:
- https://github.com/enso-org/enso/pull/6893 splits the documentation into separate pages for each method/constructor/type we have. It dramatically improves performance (less time spent updating HTML layout) and is also more readable.
- https://github.com/enso-org/enso/pull/6918. Now, we display an error message to the user when the backend reports execution failure.
- I'm finishing https://github.com/enso-org/enso/issues/6705, which will replace Tailwind with our own CSS rules and fix several styling issues (not all).

https://github.com/enso-org/enso/assets/6566674/5e7205a5-6728-4c0e-a8df-20986f30385a

**Comment by MichaelMauderer** [comment]
Still mainly working on visualization related fixes. Delayed due to some logging issues with the current IDE cloud integration. 

* Worked on #6724
* Fixed #6724

https://github.com/enso-org/enso/assets/1428930/6f2ab9ff-4cea-4740-a20a-1adf49c851f5

* Investigated #6899
[#6724](https://github.com/enso-org/enso/issues/6724)

**Comment by kazcw** [comment]
- Finished #6707, which now includes support for drawing triangles with `Rectangle`.
- #6674: I have implemented a 100-150ms improvement, this should help quite noticeably.
- I'm working on reducing the number of view invalidations. They are a root cause of slowness when handling any large graph changes, such as in #6674.

Performance investigation was partially blocked by #5813, which started happening a lot when trying to view run-graphs. I have tracked the error to corruption of the data structure `wasm-bindgen` uses to maintain Rust references to objects managed by the JS GC. I think this occurs after a `panic` leaves memory in an unspecified state, and then we re-enter Rust through an event and experience Undefined Behavior. To improve this:
- If we upgrade `wasm-bindgen`, we can enable [reference types](https://rustwasm.github.io/wasm-bindgen/reference/reference-types.html). That would eliminate this failure mode, and should be a little more efficient.
- We should cut down on the possible panics.
- Once we have reduced panics enough, we could ensure that no further code executes when a panic has occurred. This avoids the application failing in random ways when a panic has previously occurred.

---

Next:
- #6637--It might be caused by the same view-invalidation slowness as #6674.
- #6865--See how we can improve the performance of large dropdowns.

**Reply by farmaazon** [reply]
> If we upgrade wasm-bindgen, we can enable [reference types](https://rustwasm.github.io/wasm-bindgen/reference/reference-types.html). That would eliminate this failure mode, and should be a little more efficient.

As I see, we already have wasm-bindgen in newer version: https://github.com/enso-org/enso/blob/develop/Cargo.toml#L96 so I guess we could do it right now?

**Comment by mwu-tow** [comment]
Last week I worked on making all shortcut-triggered actions atomic. The fix PR #6950 should cover any synchronous operations that were previously not atomic. Notably, this includes cases like deleting connected nodes that were previously very difficult to handle correctly.

Not covered and waiting future work:
* fixes for visualizations toggling handling (it is currently race'ish): https://github.com/enso-org/enso/issues/6589;
* fixes for longer asynchronous operations, like uploading the file: https://github.com/enso-org/enso/issues/6591. This case should behave a little bit better now, but still the actual post-upload updates will create a separate transaction.

**Comment by JaroslavTulach** [comment]
Engine team focused on other things than performance:
- deadlocks - #6841
- imports - #6678
- runtime argument checks - #6682
- language server [mental experiment](https://github.com/orgs/enso-org/discussions/6762#discussioncomment-6091363)
